common/src/device/transport/mqtt/ClientWebSocketChannel.cs

```diff
@@ -242,38 +243,74 @@ protected override async void DoBeginRead()
+
+                // The parameter "direct" is used to indicate if operations carried out in the CompositeByteBuffer should be treated as "unsafe".
+                var compositeByteBuffer = new CompositeByteBuffer(allocator, direct: false, maxNumComponents: int.MaxValue);
+
+                var bytesToBeWritten = new ArraySegment<byte>();
                 _isWriteInProgress = true;
+
                 while (true)
                 {
                     object currentMessage = channelOutboundBuffer.Current;
+
+                    // Once there are no more messages pending in ChannelOutboundBuffer, the "Current" property is returned as "null".
+                    // This indicates that all pending messages have been dequeued from the ChannelOutboundBuffer and are ready to be written to the websocket.
                     if (currentMessage == null)
                     {
-                        // Wrote all messages.
+                        // This indicates that the ChannelOutboundBuffer had readable bytes and they have been added to the CompositeByteBuffer.
+                        if (compositeByteBuffer.NumComponents > 0)
+                        {
+                            // All messages have been added to the CompositeByteBuffer and are now ready to be written to the socket.
+                            bytesToBeWritten = compositeByteBuffer.GetIoBuffer();
+                        }
                         break;
                     }
 
                     var byteBuffer = currentMessage as IByteBuffer;
                     Debug.Assert(byteBuffer != null, "channelOutBoundBuffer contents must be of type IByteBuffer");
 
-                    if (byteBuffer.ReadableBytes == 0)
+                    // If the byte buffer has readable bytes then add them to the CompositeByteBuffer.
+                    if (byteBuffer.ReadableBytes != 0)
                     {
-                        channelOutboundBuffer.Remove();
-                        continue;
+                        // There are two operations carried out while adding a byte buffer component to a CompositeByteBuffer:
+                        // - Increase WriterIndex of the CompositeByteBuffer
+                        //      - increases the count of readable bytes added to the CompositeByteBuffer.
+                        // - Call the method Retain() on the byte buffer being added
+                        //      - The property ReferenceCount of a byte buffer implementation maintains a counter of the no of messages available for dequeueing.
+                        //        A ReferenceCount of 0 indicates that all messages have been flushed and the buffer can be deallocated.
+                        //        By calling the method Retain() on each byte buffer component added to the CompositeByteBuffer,
+                        //        we increase the ReferenceCount by 1 and mark them as ready for dequeuing.
```

Comment:
the spelling of dequeuing here is different than the one on line 291. Not sure which one is right :)

Reply:
probably this one?

Reply:
whoops, "dequeuing" (so many extra alphabets just sitting there needlessly)